packages/google_maps_flutter/google_maps_flutter_web/lib/src/marker_clustering.dart

```diff
@@ -0,0 +1,153 @@
+  /// emitting map events.
+  ClusterManagersController(
+      {required StreamController<MapEvent<Object?>> stream})
+      : _streamController = stream,
+        _clusterManagerIdToMarkerClusterer =
+            <ClusterManagerId, MarkerClusterer>{};
+
+  // The stream over which cluster managers broadcast their events
+  final StreamController<MapEvent<Object?>> _streamController;
+
+  // A cache of [MarkerClusterer]s indexed by their [ClusterManagerId].
+  final Map<ClusterManagerId, MarkerClusterer>
+      _clusterManagerIdToMarkerClusterer;
+
+  /// Adds a set of [ClusterManager] objects to the cache.
+  void addClusterManagers(Set<ClusterManager> clusterManagersToAdd) {
+    clusterManagersToAdd.forEach(_addClusterManager);
+  }
+
+  void _addClusterManager(ClusterManager clusterManager) {
+    final MarkerClusterer markerClusterer = createMarkerClusterer(
+        googleMap,
+        (gmaps.MapMouseEvent event, MarkerClustererCluster cluster,
+                gmaps.GMap map) =>
+            _clusterClicked(
+                clusterManager.clusterManagerId, event, cluster, map));
+
+    _clusterManagerIdToMarkerClusterer[clusterManager.clusterManagerId] =
+        markerClusterer;
+    markerClusterer.onAdd();
+  }
+
+  /// Removes a set of [ClusterManagerId]s from the cache.
+  void removeClusterManagers(Set<ClusterManagerId> clusterManagerIdsToRemove) {
+    clusterManagerIdsToRemove.forEach(_removeClusterManager);
+  }
+
+  void _removeClusterManager(ClusterManagerId clusterManagerId) {
+    final MarkerClusterer? markerClusterer =
+        _clusterManagerIdToMarkerClusterer[clusterManagerId];
+    if (markerClusterer != null) {
+      markerClusterer.clearMarkers(true);
+      markerClusterer.onRemove();
+    }
+    _clusterManagerIdToMarkerClusterer.remove(clusterManagerId);
+  }
+
+  /// Adds given [gmaps.Marker] to the [MarkerClusterer] with given
+  /// [ClusterManagerId].
+  void addItem(ClusterManagerId clusterManagerId, gmaps.Marker marker) {
+    final MarkerClusterer? markerClusterer =
+        _clusterManagerIdToMarkerClusterer[clusterManagerId];
+    if (markerClusterer != null) {
+      markerClusterer.addMarker(marker, true);
+      markerClusterer.dirty = true;
+    }
+  }
+
+  /// Removes given [gmaps.Marker] from the [MarkerClusterer] with given
+  /// [ClusterManagerId].
+  void removeItem(ClusterManagerId clusterManagerId, gmaps.Marker? marker) {
+    if (marker != null) {
+      final MarkerClusterer? markerClusterer =
+          _clusterManagerIdToMarkerClusterer[clusterManagerId];
+      if (markerClusterer != null) {
+        markerClusterer.removeMarker(marker, true);
+        markerClusterer.dirty = true;
+      }
+    }
+  }
+
+  /// Renders all cluster managers that are marked as dirty.
+  ///
+  /// This mechanism is used to batch changes to the cluster managers to avoid
+  /// unnecessary re-renders.
+  /// [MarkerClusterer] is marked as dirty when a marker is added or removed.
+  void renderDirty() {
```

Comment:
When manually testing, it seemed that some clusters disappeared, or didn't render at all after modifying them (with the example app), until I changed the zoom level. Is this function too strict? Should it re-render more?

Reply:
This doesn't seem to happen in "--profile" mode :/

Reply:
I removed the whole dirty handling, as the clustering seems to work really fast even without it and it seems like an unnecessary optimization at the moment.